### DISCUSSION TITLE
[show-and-tell] A convenience function to collect targets and refer to them by groups

### DISCUSSION THREAD

**Discussion by noamross** [ropensci/targets#1196] [Show and tell]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

This is a convenience function we've been using.  Place `all_targets()` at the end of  a `_targets.R` file and it will inspect the environment for any targets, as well as any lists that are made up of all targets (such as those created with `tar_plan()`), collect them and output them as a list.  This way you don't have to maintain a `list()` call at the bottom of the file.

We make use of `tar_plan()` a lot, both for the covenience of the `=` syntax and also to collect targets into convenient groups, like `raw_data_targets <- tar_plan(...)`.   We wanted a convenient way to build or inspect those group, so I also just added `add_list_names = TRUE` to this.  It looks for any `tar_plan()`-type lists, and after collecting their contents, writes vectors of the target names to the environment.  This way, if you have `raw_data_targets <- tar_plan(...)` in _targets.R, one can run `tar_make(raw_data_targets)`, `tar_visnetwork(TRUE, raw_data_targets)` or any other targets function that evaluates target names.  It won't work with things like `tar_load()` that don't source `_targets.R` first. 

```
#' Collect all targets and lists of targets in the environment
all_targets <- function(env = parent.env(environment()), type = "tar_target", add_list_names = TRUE) {
  
  # Function to determine if an object is a type (a target), or a list on only that type
  rfn <- function(obj) inherits(obj, type) || (is.list(obj) && all(vapply(obj, rfn, logical(1))))
  # Get the names everything in the environment (e.g. sourced in the _targets.R file)
  objs <- ls(env)
  
  out <- list()
  for(o in objs) {
    obj <- get(o, envir = env) # Get each top-level object in turn
    if (rfn(obj)) { # For targets and lists of targets
      out[[length(out) + 1]] <- obj # Add them to the output
      
      # If the object is a list of targets, add a vector of the target names to the environment
      # So that one can call `tar_make(list_name)` to make all the targets in that list
      if(add_list_names && is.list(obj)) {
        target_names <- vapply(obj, \(x) x$settings$name, character(1))
        assign(o, target_names, envir = env)
      }
    }
  }
  return(out)
}
```

Running `tar_make(raw_data_targets)` right now does give a deprecation error:

```
Using an external vector in selections was deprecated in tidyselect 1.1.0.
ℹ Please use `all_of()` or `any_of()` instead.
  # Was:
  data %>% select(raw_data_targets)

  # Now:
  data %>% select(all_of(raw_data_targets))
```

So the better way to do this is `tar_make(all_of(raw_data_targets))`

Any thoughts on this?  Is it dangerous in some way?

**Comment by wlandau** [comment]
The `names` argument of `tar_make()` operates on the names of the targets, and `tar_make()` is agnostic to any nesting structure in `_targets.R`. The idea is to set `names` equal to a `tidyselect` function call on a character vector, something like `any_of(c("target1", "target2"))` or `starts with("my_prefix_")`. So if I write:

```r
library(targets)
library(tarchetypes)
list(
  abc = tar_plan(abc1 = TRUE, xyz1 = TRUE),
  xyz = tar_plan(abc2 = TRUE, xyz2 = TRUE)
)
```

Then `tar_make(names = contains("abc"))` shows:

```r
▶ dispatched target abc1
● completed target abc1 [0 seconds]
▶ dispatched target abc2
● completed target abc2 [0 seconds]
▶ completed pipeline [0.079 seconds]
```

To interact with groups of targets in `tar_make()` and `tar_load()`, I recommend defining syntactic names that naturally express those groups.

**Reply by noamross** [reply]
Thanks Will! One of the reasons we've done this is we just found our target names were too overloaded with info to try to also add these groups as prefixes.  Also we wanted to avoid re-naming things in a pipeline with a lot of large/long-compute targets.
